src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java

```diff
@@ -1243,6 +1246,24 @@ public ExecGroupCollection getExecGroups() {
     return execGroupCollection;
   }
 
+  private final static ConstraintSettingInfo osConstraintSetting;
+  private final static ConstraintValueInfo windowsConstraintValue;
+
+  static {
+    try {
+      Label osConstraintSettingLabel = Label.parseAbsolute("@platforms//os:os", ALWAYS_FALLBACK);
```

Comment:
This will probably fail with Bzlmod as it doesn't use a repository mapping - with Bzlmod, the `platforms` module will manifest as a repository named something like `@platforms~0.0.4`.

You should be able to use:
 
```
Label.parseWithRepoContext("@platforms//os:os", repoContext.of(RepositoryName.MAIN, getRule.getPackage().getRepositoryMapping()));
```
but that won't work from a static context.

Reply:
Thanks. Should I just parse the labels directly in `isTargetOsWindows`?

Reply:
I unfortunately don't know enough about constraint handling in Java code to answer that question - I would leave it for the reviewer.

Reply:
Thinking about this some more, it's more involved and what I suggested above is incorrect: The rule may be instantiated in *any* repository, not necessarily the main repository. But these repositories may not depend on the `platforms` module and thus may not even see `@platforms` in their repository mapping.

@Wyverald can probably help here, but may still be OOO.

Reply:
@Wyverald Do you happen to know the best way to test for the `@platforms//os:windows` constraint on the target platform inside a Java rule implementation?

Reply:
I noticed the way to use `@platforms//os:windows` in a Starlark rule is to express a dependency on it which allows pulling out a `ConstraintValueInfo` provider in the rule implementation to pass to `target_platform_has_constraint`: https://github.com/aspect-build/bazel-lib/blob/5edb10a0d8d207ba9379322bcebb58f541dc8060/lib/private/copy_file.bzl#L116

Is the right thing to do something similar on the Java side?

Reply:
@comius would be the best person to answer this, I think. I really don't know whether we're supposed to grab labels like this in Java code.

With that said, what @fmeum pointed out still stands -- we can't rely on the current repository having visibility into the `@platforms` repo. Depending on the answer to the previous question, we might have ways to work around this.

Reply:
Please move all references to Windows out of RuleContext. 

RuleContext should remain a general device, and is no place for specifics. You can either put the code into into each rule or providing a helper class under `rules` or `bazel/rules` package.

This will allso make it possible to Starlarkify the rule's your modifying. There's already available API in Starlark `ctx.target_platform_has_constraint`.

Creating a providers instead of using an implicit dependency to `@platforms//os:windows` is also not rewritable to Starlark (you can't create a ConstraintInfoProvider). Please use the example you found in copy_file.

Reply:
@comius Thanks. How do I get a `Label` for `@platforms//os:windows` to use in `Bazel{Sh,Py,Java}BinaryRule.build`? It looks like `RuleDefinitionEnvironment` can only give Labels for things in `@bazel_tools`. Will this work?

```java
.add(
    attr("_windows_constraint", LABEL)
        .value(Label.parseAbsoluteUnchecked("@platforms//os:windows")))
```

Reply:
the above will likely not work with Bzlmod (same reason as @fmeum described earlier). `@bazel_tools` is special since it's a "built-in module" which is shipped with the Bazel binary; `@platforms` is not.

You would be able to use labels from `@platforms` if you were defining these rules in Starlark (in the builtins_bzl stuff, for example). Not sure where we are on that front -- I think `java_binary` is already Starlarkified?

Reply:
It think this would work, except if I'm missing some bzlmod machinery that is invoked on Labels in Starlark.

There's another problem: it will only work in Bazel, not in a monorepo. We use getToolsRepository to map it to absolute label. Can you extend it with `getPlatfromsRepository()`? This should make it possible to remap @platforms as well.